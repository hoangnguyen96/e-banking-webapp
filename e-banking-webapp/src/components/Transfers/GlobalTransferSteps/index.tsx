'use client';

// Libs
import { useForm } from 'react-hook-form';
import { z } from 'zod';
import { zodResolver } from '@hookform/resolvers/zod';
import { Session } from 'next-auth';

// Models
import { IAccountPayloadData, TransactionCreateData } from '@/interfaces';

// API
import { createTransaction, updateAccountInfo } from '@/actions';

// Helpers / Utils
import { GlobalTransferFormSchema } from '@/schemas';
import { convertToUSD, formatNumberWithCommas } from '@/utils';

// Components
import * as WizardForm from '@/components/common/WizardForm';
import {
  GlobalTransferForm,
  ConfirmGlobalTransfer,
  GlobalTransferSuccess,
} from '@/components';

type FormValues = z.infer<typeof GlobalTransferFormSchema>;

interface IGlobalTransferSteps {
  session: Session;
  onClose?: () => void;
}

export const GlobalTransferSteps = ({
  session,
  onClose,
}: IGlobalTransferSteps) => {
  const form = useForm<FormValues>({
    mode: 'onBlur',
    defaultValues: {
      fromAccountType: undefined,
      fromCountryType: undefined,
      amount: 0,
    },
    resolver: zodResolver(GlobalTransferFormSchema),
  });

  const allFieldValues = form.watch();

  const amountInUSD = formatNumberWithCommas(
    convertToUSD(allFieldValues.fromCountryType, allFieldValues.amount),
  );

  const submitHandler = async ({
    fromAccountId,
    fromCardName,
    fromAccountNumber,
    fromAccountBalance,
    fromAccountType,
    amount,
  }: FormValues) => {
    const transactionData: TransactionCreateData = {
      fromAccountId,
      toAccountId: allFieldValues.recipientAccount,
      fromAccountType,
      toAccountType: undefined,
      statusTransaction: true,
      amount: Number(amount),
    };

    const payload: IAccountPayloadData = {
      name: fromCardName,
      accountNumber: fromAccountNumber,
      balance: fromAccountBalance - Number(amount),
      type: fromAccountType,
      currency: '$',
    };

    await createTransaction(transactionData);
    await updateAccountInfo(fromAccountId, payload);
  };

  return (
    <WizardForm.Root
      schema={GlobalTransferFormSchema}
      form={form}
      className='flex grow flex-col'
    >
      <WizardForm.Step name='transfer' key='transfer'>
        <GlobalTransferForm session={session} />
      </WizardForm.Step>
      <WizardForm.Step
        name='confirm'
        key='confirm'
        className='flex grow flex-col items-center justify-center'
      >
        <ConfirmGlobalTransfer
          {...allFieldValues}
          submitHandler={submitHandler}
          amountInUSD={amountInUSD}
          userName='Yehudi Daud'
        />
      </WizardForm.Step>
      <WizardForm.Step
        name='success'
        key='success'
        className='flex grow flex-col items-center justify-center'
      >
        <GlobalTransferSuccess
          {...allFieldValues}
          onClose={onClose}
          userName='Yehudi Daud'
        />
      </WizardForm.Step>
    </WizardForm.Root>
  );
};
