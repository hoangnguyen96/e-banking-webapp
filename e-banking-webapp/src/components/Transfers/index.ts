export * from './InternalTransferForm';
export * from './ConfirmInternalTransfer';
export * from './ConfirmGlobalTransfer';
export * from './InternalTransferSuccess';
export * from './GlobalTransferForm';
