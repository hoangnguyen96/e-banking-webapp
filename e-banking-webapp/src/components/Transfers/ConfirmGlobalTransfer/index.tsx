'use client';

// Interfaces
import { IAccount, AccountType, CurrencyUnit } from '@/interfaces';

// Components
import { Button, Text } from '@/components';

// Utils
import { formatNumberWithCommas } from '@/utils';

interface IConfirmGlobalTransfer {
  amount: number;
  currencyUnit?: CurrencyUnit;
  fromAccountType: AccountType;
  userName: IAccount['name'];
  onCancel: () => void;
  onProceed: () => void;
}

export const ConfirmGlobalTransfer = ({
  amount,
  currencyUnit = '$',
  userName,
  fromAccountType,
  onCancel,
  onProceed,
}: IConfirmGlobalTransfer) => (
  <div className='flex w-[252px] flex-col items-center text-center'>
    <Text className='text-sm font-semibold text-navyBlue'>
      {userName} is about to receive
    </Text>
    <Text className='mt-3 text-4xl font-bold text-navyBlue'>
      {currencyUnit} {formatNumberWithCommas(amount)}
    </Text>
    <Text className='mt-3 text-xs font-medium text-transparentBlack'>
      From your {fromAccountType} wallet to your {userName} wallet, this action
      cannot be undone once approved...
    </Text>
    <div className='mt-10 flex gap-6'>
      <Button radius='xs' color='danger' size='xxl' onClick={onCancel}>
        Cancel
      </Button>
      <Button radius='xs' color='navyBlue' size='xxl' onClick={onProceed}>
        Procced
      </Button>
    </div>
  </div>
);
