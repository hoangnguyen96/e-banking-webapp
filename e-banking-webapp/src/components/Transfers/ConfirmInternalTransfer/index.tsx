'use client';

// Interfaces
import { AccountType, CurrencyUnit } from '@/interfaces';

// Components
import { Button, Text } from '@/components';

// Utils
import { formatNumberWithCommas } from '@/utils';

interface ConfirmInternalTransferProps {
  amount: number;
  currencyUnit?: CurrencyUnit;
  fromAccountType: AccountType;
  toAccountType: AccountType;
  onCancel: () => void;
  onConfirm: () => void;
}

export const ConfirmInternalTransfer = ({
  amount,
  currencyUnit = '$',
  fromAccountType,
  toAccountType,
  onCancel,
  onConfirm,
}: ConfirmInternalTransferProps) => {
  return (
    <div className='flex flex-col items-center justify-center gap-4'>
      <Text as='h4' className='text-sm font-semibold'>
        You are about to transfer
      </Text>
      <Text as='span' className='text-4xl font-bold'>
        {currencyUnit} {formatNumberWithCommas(amount)}
      </Text>
      <Text className='max-w-[250px] text-center text-xs font-medium text-primary-200 opacity-50'>
        From your {fromAccountType} wallet to your {toAccountType} wallet, this
        action cannot be undone once approved...
      </Text>
      <div className='mt-8 flex gap-8'>
        <Button color='tertiary' radius='xs' size='md' onClick={onCancel}>
          Cancel
        </Button>
        <Button color='primary' radius='xs' size='md' onClick={onConfirm}>
          Proceed
        </Button>
      </div>
    </div>
  );
};
