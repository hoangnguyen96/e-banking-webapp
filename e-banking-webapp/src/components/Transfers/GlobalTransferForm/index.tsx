'use client';

// Third Party
import { Controller, useWatch } from 'react-hook-form';
import { Session } from 'next-auth';
import { z } from 'zod';
import { useEffect, useState, useTransition } from 'react';
import { Spinner } from '@nextui-org/react';

// Constants
import {
  ERROR_MESSAGES,
  OPTIONS_COUNTRY_CODE_CONVERT_GLOBAL,
  TRANSFER_FORM_ACCOUNT_OPTIONS,
  TRANSFER_FORM_GLOBAL_OPTIONS,
} from '@/constants';

// Interfaces
import { AccountType, GlobalType } from '@/interfaces';

// API
import { getAccountInfoByAccountType } from '@/services';

// Helpers / Utils
import {
  convertToUSD,
  formatNumberWithCommas,
  isValidNumber,
  sanitizeNumber,
} from '@/utils';
import { GlobalTransferFormSchema } from '@/schemas';

// Components
import { Button, Input, Select, Text, SendIcon } from '@/components';

// Contexts
import { useWizardFormContext } from '@/context';

type FormValues = keyof z.infer<typeof GlobalTransferFormSchema>;

export const GlobalTransferForm = ({ session }: { session: Session }) => {
  const {
    form: {
      control,
      formState: { errors },
      setValue,
      getValues,
      setError,
      clearErrors,
    },
    onNextStep,
    validateStep,
  } = useWizardFormContext<typeof GlobalTransferFormSchema>();

  const hiddenFields: FormValues[] = [
    'fromAccountId',
    'fromCardName',
    'fromAccountNumber',
    'fromAccountBalance',
  ];

  const fromAccountTypeValue = useWatch({
    control,
    name: 'fromAccountType',
  });

  const fromCountryType = useWatch({
    control,
    name: 'fromCountryType',
  });

  const allFieldValues = getValues();

  const amountValueInUSD = convertToUSD(
    allFieldValues.fromCountryType,
    allFieldValues.amount,
  );

  const [isPending, startTransition] = useTransition();

  // States for balances
  const [balanceSend, setBalanceSend] = useState<number | null>(null);
  const [rawAmount, setRawAmount] = useState<string>('');

  useEffect(() => {
    const fetchBalanceSend = async () => {
      if (!fromAccountTypeValue) {
        setBalanceSend(null);
        return;
      }

      startTransition(async () => {
        try {
          const balance = await getAccountInfoByAccountType(
            session.user.id,
            fromAccountTypeValue,
            'balance',
          );

          const accountId = await getAccountInfoByAccountType(
            session.user.id,
            fromAccountTypeValue,
            'documentId',
          );

          const fromCardName = await getAccountInfoByAccountType(
            session.user.id,
            fromAccountTypeValue,
            'name',
          );

          const fromAccountNumber = await getAccountInfoByAccountType(
            session.user.id,
            fromAccountTypeValue,
            'accountNumber',
          );

          setBalanceSend(Number(balance));
          setValue('fromAccountId', String(accountId));
          setValue('fromCardName', String(fromCardName));
          setValue('fromAccountNumber', String(fromAccountNumber));
          setValue('fromAccountBalance', Number(balance));
        } catch (error) {
          console.error('Error fetching balance for send account:', error);
        }
      });
    };

    fetchBalanceSend();
  }, [fromAccountTypeValue, session.user.id, setValue, startTransition]);

  const countryCode = () => {
    return (
      OPTIONS_COUNTRY_CODE_CONVERT_GLOBAL.find(
        (option) => option.key === fromCountryType,
      )?.label || ''
    );
  };

  const filteredFromAccountOptions = () =>
    TRANSFER_FORM_ACCOUNT_OPTIONS.filter(
      (option) => option.key !== (fromCountryType as unknown as AccountType),
    );

  const filteredToAccountOptions = () =>
    TRANSFER_FORM_GLOBAL_OPTIONS.filter(
      (option) =>
        option.key !== (fromAccountTypeValue as unknown as GlobalType),
    );

  const handleInputChange = (
    value: string,
    onChange: (value: string) => void,
  ) => {
    const sanitizedValue = sanitizeNumber(value);
    if (isValidNumber(sanitizedValue)) {
      setRawAmount(sanitizedValue);
      onChange(sanitizedValue);
    }
  };

  return (
    <div className='flex w-full flex-col gap-4'>
      {/* Title */}
      <Text as='h4' className='text-xs font-medium'>
        Send Money Across The Global with Ease
      </Text>

      {/* Country */}
      <Controller
        control={control}
        name='fromCountryType'
        render={({ field: { onChange, onBlur, value } }) => {
          return (
            <Select
              label='Country'
              labelPlacement='outside'
              options={filteredToAccountOptions()}
              classNames={{ label: 'text-sm' }}
              value={String(value)}
              errorMessage={errors.fromCountryType?.message}
              isInvalid={!!errors.fromCountryType}
              selectedKeys={value ? [String(value)] : []}
              onSelectionChange={(keys) => {
                const selectedValue = String(Array.from(keys)[0]);
                onChange(selectedValue);
              }}
              onClose={onBlur}
            />
          );
        }}
      />

      {/* Account */}
      <Controller
        control={control}
        name='fromAccountType'
        render={({ field: { onChange, onBlur, value } }) => {
          return (
            <Select
              label='Account'
              labelPlacement='outside'
              options={filteredFromAccountOptions()}
              classNames={{ label: 'text-sm' }}
              value={String(value)}
              errorMessage={errors.fromAccountType?.message}
              isInvalid={!!errors.fromAccountType}
              isDisabled={isPending}
              selectedKeys={value ? [String(value)] : []}
              onSelectionChange={(keys) => {
                const selectedValue = String(Array.from(keys)[0]);
                onChange(selectedValue);
              }}
              onClose={onBlur}
            />
          );
        }}
      />

      {/* Recipient Account */}
      {fromAccountTypeValue && (
        <Controller
          control={control}
          name='recipientAccount'
          render={({ field: { value, onChange, onBlur } }) => {
            return (
              <Input
                label='Recipient Account'
                labelPlacement='outside'
                placeholder=' '
                classNames={{
                  inputWrapper:
                    'h-10 px-2.5 py-2 rounded-sm border-default box-border',
                  input: 'm-0 text-xs text-primary-200 font-medium',
                }}
                value={value}
                errorMessage={errors.recipientAccount?.message}
                isInvalid={!!errors.recipientAccount}
                onChange={onChange}
                onBlur={onBlur}
                maxLength={12}
              />
            );
          }}
        />
      )}

      {/* Available Balance */}
      <div className='flex items-center gap-2'>
        <Text as='span' className='text-xs text-foreground-200 opacity-50'>
          Available Balance:
        </Text>
        <Text as='span' className='text-xs text-foreground-200 opacity-50'>
          {isPending ? (
            <Spinner size='sm' />
          ) : (
            balanceSend && `$${formatNumberWithCommas(balanceSend)}`
          )}
        </Text>
      </div>

      {/* Amount */}
      <Controller
        control={control}
        name='amount'
        render={({ field: { value, onChange } }) => {
          return (
            <div className='flex items-baseline gap-[15px]'>
              <Input
                value={countryCode()}
                readOnly
                classNames={{
                  base: 'w-11',
                  inputWrapper:
                    'w-11 h-10 rounded-sm border-default box-border p-0',
                  input: 'font-semibold text-navyBlue text-xs text-center',
                }}
              />
              <Input
                inputMode='decimal'
                label='Amount'
                labelPlacement='outside'
                placeholder=' '
                classNames={{
                  inputWrapper:
                    'h-10 px-2.5 py-2 rounded-sm border-default box-border',
                  input: 'm-0 text-xs text-primary-200 font-medium',
                }}
                value={
                  rawAmount
                    ? formatNumberWithCommas(Number(rawAmount))
                    : formatNumberWithCommas(value)
                }
                onChange={(e) => handleInputChange(e.target.value, onChange)}
                errorMessage={errors.amount?.message}
                isInvalid={!!errors.amount}
                onBlur={() => {
                  if (balanceSend) {
                    if (amountValueInUSD > balanceSend) {
                      setError('amount', {
                        type: 'validate',
                        message: ERROR_MESSAGES.AMOUNT_EXCEEDED_BALANCE,
                      });
                    } else {
                      clearErrors('amount');
                    }
                  }
                }}
              />
            </div>
          );
        }}
      />

      {/* Hidden Fields */}
      {hiddenFields.map((fieldName) => (
        <Controller
          key={fieldName}
          control={control}
          name={fieldName}
          render={({ field: { value } }) => (
            <Input value={String(value)} className='hidden' />
          )}
        />
      ))}

      <Button
        type='submit'
        startContent={<SendIcon />}
        className='bg-primary-200 font-semibold text-foreground-200'
        onClick={onNextStep}
        isDisabled={!validateStep()}
      >
        Transfer Funds
      </Button>
    </div>
  );
};
