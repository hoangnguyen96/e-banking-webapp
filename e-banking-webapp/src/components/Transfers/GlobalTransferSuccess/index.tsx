'use client';

// Interface
import { CurrencyUnit, IAccount } from '@/interfaces';

// Utils
import { formatNumberWithCommas } from '@/utils';

// Components
import { Button, CheckIcon, Text } from '@/components';

interface IGlobalTransferSuccess {
  amount: number;
  currencyUnit?: CurrencyUnit;
  userName: IAccount['name'];
  onClose: () => void;
}

export const GlobalTransferSuccess = ({
  amount,
  currencyUnit = '$',
  userName,
  onClose,
}: IGlobalTransferSuccess) => (
  <div className='flex flex-col items-center gap-4'>
    <div className='flex h-14 w-14 items-center justify-center rounded-full bg-background-300'>
      <CheckIcon />
    </div>
    <Text as='h4' className='text-sm font-semibold'>
      Transaction was Successful
    </Text>
    <Text className='max-w-[220px] text-center text-xs font-medium text-primary-200 opacity-50'>
      A sum of {currencyUnit} {formatNumberWithCommas(amount)} was transferred
      to {userName}
    </Text>

    <Button
      color='tertiary'
      radius='xs'
      size='md'
      className='mt-8'
      onClick={onClose}
    >
      Close
    </Button>
  </div>
);
