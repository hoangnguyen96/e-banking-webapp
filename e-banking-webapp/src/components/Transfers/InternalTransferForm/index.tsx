'use client';

// Libs
import { z } from 'zod';
import { useEffect, useState } from 'react';
import { Controller, useWatch } from 'react-hook-form';
import { Spinner } from '@nextui-org/react';

import { Session } from 'next-auth';

// Constants
import { TRANSFER_FORM_ACCOUNT_OPTIONS } from '@/constants';

// Components
import { Button, Input, Select, SendIcon, Text } from '@/components';

// Schemas
import { InternalTransferFormSchema } from '@/schemas';

// Context
import { useWizardFormContext } from '@/context';

// Services
import { getAccountInfoByAccountType } from '@/services';

// Utils
import { formatNumberWithCommas } from '@/utils';

type FormValues = keyof z.infer<typeof InternalTransferFormSchema>;
interface IInternalTransferFormProps {
  session: Session;
}

export const InternalTransferForm = ({
  session,
}: IInternalTransferFormProps) => {
  const {
    form: {
      control,
      formState: { errors },
      setValue,
    },
    onNextStep,
    validateStep,
  } = useWizardFormContext<typeof InternalTransferFormSchema>();

  const hiddenFields: FormValues[] = [
    'fromAccountId',
    'toAccountId',
    'fromCardName',
    'toCardName',
    'fromAccountNumber',
    'toAccountNumber',
    'fromAccountBalance',
    'toAccountBalance',
  ];

  const fromAccountTypeValue = useWatch({
    control,
    name: 'fromAccountType',
  });

  const toAccountTypeValue = useWatch({
    control,
    name: 'toAccountType',
  });

  // States for fetching
  const [isFetchingBalanceSend, setIsFetchingBalanceSend] = useState(false);
  const [isFetchingBalanceReceive, setIsFetchingBalanceReceive] =
    useState(false);

  // States for balances
  const [balanceSend, setBalanceSend] = useState<number | null>(null);
  const [balanceReceive, setBalanceReceive] = useState<number | null>(null);

  useEffect(() => {
    const fetchBalanceSend = async () => {
      if (!fromAccountTypeValue) {
        setBalanceSend(null);
        return;
      }

      try {
        setIsFetchingBalanceSend(true);

        const balance = await getAccountInfoByAccountType(
          session.user.id,
          fromAccountTypeValue,
          'balance',
        );

        const accountId = await getAccountInfoByAccountType(
          session.user.id,
          fromAccountTypeValue,
          'documentId',
        );

        const fromCardName = await getAccountInfoByAccountType(
          session.user.id,
          fromAccountTypeValue,
          'name',
        );

        const fromAccountNumber = await getAccountInfoByAccountType(
          session.user.id,
          fromAccountTypeValue,
          'accountNumber',
        );

        setBalanceSend(Number(balance));
        setValue('fromAccountId', String(accountId));
        setValue('fromCardName', String(fromCardName));
        setValue('fromAccountNumber', String(fromAccountNumber));
        setValue('fromAccountBalance', Number(balance));
      } catch (error) {
        console.error('Error fetching balance for send account:', error);
      } finally {
        setIsFetchingBalanceSend(false);
      }
    };

    fetchBalanceSend();
  }, [fromAccountTypeValue, session.user.id, setValue]);

  useEffect(() => {
    const fetchBalanceReceive = async () => {
      if (!toAccountTypeValue) {
        setBalanceReceive(null);
        return;
      }
      try {
        setIsFetchingBalanceReceive(true);

        const balance = await getAccountInfoByAccountType(
          session.user.id,
          toAccountTypeValue,
          'balance',
        );

        const documentId = await getAccountInfoByAccountType(
          session.user.id,
          toAccountTypeValue,
          'documentId',
        );

        const toCardName = await getAccountInfoByAccountType(
          session.user.id,
          toAccountTypeValue,
          'name',
        );

        const toAccountNumber = await getAccountInfoByAccountType(
          session.user.id,
          toAccountTypeValue,
          'accountNumber',
        );

        setBalanceReceive(Number(balance));
        setValue('toAccountId', String(documentId));
        setValue('toCardName', String(toCardName));
        setValue('toAccountNumber', String(toAccountNumber));
        setValue('toAccountBalance', Number(balance));
      } catch (error) {
        console.error('Error fetching balance for receive account:', error);
      } finally {
        setIsFetchingBalanceReceive(false);
      }
    };

    fetchBalanceReceive();
  }, [toAccountTypeValue, session.user.id, setValue]);

  return (
    <form className='flex flex-col gap-4'>
      {/* Title */}
      <Text as='h4' className='text-xs font-medium'>
        Send Money Across Your eWallet Accounts
      </Text>

      {/* Account receive */}
      <Controller
        control={control}
        name='toAccountType'
        render={({ field: { onChange, onBlur, value } }) => {
          const filteredOptions = TRANSFER_FORM_ACCOUNT_OPTIONS.filter(
            (option) => option.key !== fromAccountTypeValue,
          );

          return (
            <>
              <Select
                label='To'
                labelPlacement='outside'
                options={filteredOptions}
                classNames={{ label: 'text-sm' }}
                value={String(value)}
                errorMessage={errors.toAccountType?.message}
                isInvalid={!!errors.toAccountType}
                onSelectionChange={(keys) => {
                  const selectedValue = String(Array.from(keys)[0]);
                  onChange(selectedValue);
                }}
                onClose={onBlur}
              />
              <div className='flex items-center gap-2'>
                <Text
                  as='span'
                  className='text-xs text-foreground-200 opacity-50'
                >
                  Available Balance:
                </Text>
                <Text
                  as='span'
                  className='text-xs text-foreground-200 opacity-50'
                >
                  {isFetchingBalanceReceive ? (
                    <Spinner size='sm' />
                  ) : (
                    balanceReceive &&
                    `$${formatNumberWithCommas(balanceReceive)}`
                  )}
                </Text>
              </div>
            </>
          );
        }}
      />

      {/* Account send */}
      <Controller
        control={control}
        name='fromAccountType'
        render={({ field: { onChange, onBlur, value } }) => {
          const filteredOptions = TRANSFER_FORM_ACCOUNT_OPTIONS.filter(
            (option) => option.key !== toAccountTypeValue,
          );

          return (
            <>
              <Select
                label='From'
                labelPlacement='outside'
                options={filteredOptions}
                classNames={{ label: 'text-sm' }}
                value={String(value)}
                errorMessage={errors.fromAccountType?.message}
                isInvalid={!!errors.fromAccountType}
                onSelectionChange={(keys) => {
                  const selectedValue = String(Array.from(keys)[0]);
                  onChange(selectedValue);
                }}
                onClose={onBlur}
              />
              <div className='flex items-center gap-2'>
                <Text
                  as='span'
                  className='text-xs text-foreground-200 opacity-50'
                >
                  Available Balance:
                </Text>
                <Text
                  as='span'
                  className='text-xs text-foreground-200 opacity-50'
                >
                  {isFetchingBalanceSend ? (
                    <Spinner size='sm' />
                  ) : (
                    balanceSend && `$${formatNumberWithCommas(balanceSend)}`
                  )}
                </Text>
              </div>
            </>
          );
        }}
      />

      {/* Amount */}
      <Controller
        control={control}
        name='amount'
        render={({ field: { onChange, onBlur, value } }) => (
          <Input
            inputMode='decimal'
            label='Amount'
            labelPlacement='outside'
            placeholder=' '
            classNames={{
              inputWrapper: 'px-2.5 py-2 rounded-sm border-default',
              input: 'm-0 text-xs text-primary-200 font-medium',
            }}
            value={String(value)}
            errorMessage={errors.amount?.message}
            isInvalid={!!errors.amount}
            onChange={onChange}
            onBlur={onBlur}
          />
        )}
      />

      {/* Hidden Fields */}
      {hiddenFields.map((fieldName) => (
        <Controller
          key={fieldName}
          control={control}
          name={fieldName}
          render={({ field: { value } }) => (
            <Input value={String(value)} className='hidden' />
          )}
        />
      ))}

      <Button
        type='submit'
        startContent={<SendIcon />}
        className='bg-primary-200 font-semibold text-foreground-200'
        isDisabled={!validateStep()}
        onClick={onNextStep}
      >
        Transfer Funds
      </Button>
    </form>
  );
};
