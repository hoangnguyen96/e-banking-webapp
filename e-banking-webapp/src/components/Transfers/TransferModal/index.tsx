'use client';

// Constants
import { TRANSFER_TABS } from '@/constants';

// Components
import { TransferTab, Text, Modal } from '@/components';

interface ITransferModalProps {
  isOpen: boolean;
  onClose: () => void;
}

export const TransferModal = ({ isOpen, onClose }: ITransferModalProps) => {
  return (
    <Modal
      isOpen={isOpen}
      onClose={onClose}
      size='4xl'
      placement='center'
      classNames={{ base: 'p-0 rounded-xl' }}
    >
      <div className='absolute left-4 top-4 flex flex-col gap-4'>
        <Text as='h4' className='text-base font-medium text-primary-200'>
          Good Evening,{' '}
          <Text as='span' className='font-normal'>
            Pheroxios
          </Text>
        </Text>
        <Text as='h5' className='text-sm font-semibold text-primary-200'>
          Transfer Funds
        </Text>
      </div>

      <TransferTab
        TransferTabs={TRANSFER_TABS}
        classNames={{
          base: 'px-4 py-24 bg-lightGraySolid flex-1 rounded-xl min-w-[460px]',
          tabList: 'p-0',
          panel: 'py-8 pl-6 pr-8 min-h-[560px] bg-background-500 rounded-xl',
        }}
      />
    </Modal>
  );
};
