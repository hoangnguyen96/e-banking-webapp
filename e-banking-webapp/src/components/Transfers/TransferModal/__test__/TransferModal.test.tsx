// Libs
import { render } from '@testing-library/react';
import { TransferModal } from '@/components/Transfers/TransferModal';
import { useSession } from 'next-auth/react';

jest.mock('next-auth/react', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue({}),
  useSession: jest.fn(),
}));

describe('TransferModal component', () => {
  const mockUseSession = useSession as jest.Mock;
  it('should match snapshot for TransferModal', () => {
    mockUseSession.mockReturnValue({
      data: {
        user: { name: 'admin' },
        expires: '2024-01-01T00:00:00.000Z',
      },
      status: 'authenticated',
    });

    const { container } = render(
      <TransferModal isOpen={true} onClose={() => {}} />,
    );

    expect(container).toMatchSnapshot();
  });
});
