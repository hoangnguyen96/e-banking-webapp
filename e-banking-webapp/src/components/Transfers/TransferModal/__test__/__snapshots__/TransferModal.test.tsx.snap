// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransferModal component should match snapshot for TransferModal 1`] = `
<div
  aria-hidden="true"
/>
`;
